File: leetcode/leetcode_py/longest-common-prefix.py
'''
https://leetcode.com/problems/longest-common-prefix

Solution1. Iterate. find window
time: o(n)
space: o(1)

'''
class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        if len(strs) == 0: return ""
        res = strs[0]
        r = len(res)
        for i in range(1, len(strs)):
            s = strs[i]
            r = min(r, len(s))
            j = 0
            while (j < len(s) and j < r):
                if res[j] != s[j]:
                    if j == 0: return ""
                    r = j
                    break
                j += 1
        return res[:r]
        

File: leetcode/leetcode_cpp/longest-common-prefix.cpp
/*
https://leetcode.com/problems/longest-common-prefix

Solution1. Iterate. find window
time: o(n)
space: o(1)

*/
class Solution {
public:
    string longestCommonPrefix(vector<string>& strs) {
        if (!strs.size()) return "";
        auto& res = strs[0];
        int r = res.size();
        for (int i=1; i<strs.size(); ++i) {
            auto&s = strs[i];
            r = std::min(r, (int)s.size());
            for (int j=0; j<s.size() && j<r; j++) {
                if (res[j] != s[j]) {
                    if (j == 0) return "";
                    r = j;
                    break;
                }
            }
        }
        return res.substr(0, r);
    }
};